tests/hazmat/primitives/test_argon2.py

```diff
@@ -0,0 +1,163 @@
+        assert binascii.hexlify(argon2id.derive(password)) == derived_key
+
+    def test_invalid_types(self, backend):
+        with pytest.raises(TypeError):
+            Argon2id(
+                salt="notbytes",  # type: ignore[arg-type]
+                length=32,
+                iterations=1,
+                lanes=1,
+                memory_cost=32,
+                ad=None,
+                secret=None,
+            )
+
+        with pytest.raises(TypeError):
+            Argon2id(
+                salt=b"b" * 8,
+                length=32,
+                iterations=1,
+                lanes=1,
+                memory_cost=32,
+                ad="string",  # type: ignore[arg-type]
+                secret=None,
+            )
+
+        with pytest.raises(TypeError):
+            Argon2id(
+                salt=b"b" * 8,
+                length=32,
+                iterations=1,
+                lanes=1,
+                memory_cost=32,
+                ad=None,
+                secret="string",  # type: ignore[arg-type]
+            )
+
+    @pytest.mark.parametrize(
+        "params",
+        [
+            (b"b" * 8, "notanint", 1, 1, 32),  # invalid type
+            (b"b" * 8, 32, "notanint", 1, 32),  # invalid type
+            (b"b" * 8, 32, 1, "notanint", 32),  # invalid type
+            (b"b" * 8, 32, 1, 1, "notanint"),  # invalid type
+            (b"b" * 7, 3, 1, 1, 32),  # salt < 8
+            (b"b" * 8, 3, 1, 1, 32),  # length < 4
+            (b"b" * 8, 32, 0, 1, 32),  # iterations < 1
+            (b"b" * 8, 32, 1, 0, 32),  # lanes < 1
+            (b"b" * 8, 32, 1, 1, 7),  # memory_cost < 8 * lanes
+            (b"b" * 8, 32, 1, 32, 200),  # memory_cost < 8 * lanes
+        ],
+    )
+    def test_invalid_values(self, params, backend):
+        (salt, length, iterations, lanes, memory_cost) = params
+        with pytest.raises(ValueError):
+            Argon2id(
+                salt=salt,
+                length=length,
+                iterations=iterations,
+                lanes=lanes,
+                memory_cost=memory_cost,
+            )
+
+    def test_already_finalized(self, backend):
+        argon2id = Argon2id(
+            salt=b"salt" * 2, length=32, iterations=1, lanes=1, memory_cost=32
+        )
+        argon2id.derive(b"password")
+        with pytest.raises(AlreadyFinalized):
+            argon2id.derive(b"password")
+
+    def test_already_finalized_verify(self, backend):
+        argon2id = Argon2id(
+            salt=b"salt" * 2, length=32, iterations=1, lanes=1, memory_cost=32
+        )
+        digest = argon2id.derive(b"password")
+        with pytest.raises(AlreadyFinalized):
+            argon2id.verify(b"password", digest)
+
+    def test_invalid_verify(self, backend):
+        argon2id = Argon2id(
+            salt=b"salt" * 2, length=32, iterations=1, lanes=1, memory_cost=32
+        )
+        with pytest.raises(InvalidKey):
+            argon2id.verify(b"password", b"invalidkey")
```

Comment:
There should probably be a testcase where the key is the right length, but wrong value